### PR TITLE
Fix e-notices on Address settings page, convert YesNo field

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
This fixes enotices 

Before
----------------------------------------
screenshot 2018-10-25 09 38 22

After
----------------------------------------
screenshot 2018-10-25 10 04 18

Technical Details
----------------------------------------
We are converging the logic for handling settings on settings form into the SettingTrait - which is used by both Settings forms & Preferences. The difference between settings forms & preferences is arbitrary & historical & we are working to eliminate it. The final trait will be usable by extensions too. In fixing this e-notice I am also converting one of the fields over

@mattwire in doing this I wanted to look at your proposed changes here

https://github.com/civicrm/civicrm-core/pull/12932/files#diff-cd9367d6c33a74944159c2f7623db4a7R189 where you suggest changing html_type from 'radio' to 'YesNo' - my take on this is that the Quickform type is a quickform specific thing (like YesNo) but the html_type is a generic html type of element - I note this is NOT clearly documented & the docs treat html_type as a sub element of QF type. I suggest we discuss here & agree & then add a list of valid options here https://docs.civicrm.org/dev/en/latest/framework/setting/

Also I note that with the SettingTrait currently gives preference to quick_form_type & then handles html_type

```
  /**
   * Get the quickform type for the given html type.
   *
   * @param array $spec
   *
   * @return string
   */
  protected function getQuickFormType($spec) {
    if (isset($spec['quick_form_type'])) {
      return $spec['quick_form_type'];
    }
    $mapping = [
      'checkboxes' => 'CheckBoxes',
      'checkbox' => 'CheckBox',
      'radio' => 'Radio',
      'select' => 'Select',
      'textarea' => 'Element',
    ];
    return $mapping[$spec['html_type']];
  }
```

Comments
----------------------------------------
I don't think the discussion needs to block merge here - I think this is mergeable & discussion can continue

### CODE REVIEW

**Comment by mattwire** [comment]
@eileenmcnaughton Needs rebase (conflict)

File: CRM/Admin/Form/SettingTrait.php [hunk 1]
```diff
@@ -171,9 +171,13 @@ protected function addFieldsDefinedInSettingsMetadata() {
         elseif ($add == 'addMonthDay') {
           $this->add('date', $setting, ts($props['title']), CRM_Core_SelectValues::date(NULL, 'M d'));
         }
+        elseif ($add === 'addYesNo' && ($props['type'] === 'Boolean')) {
+          $this->addRadio($setting, ts($props['title']), array(0 => 'No', 1 => 'Yes'), NULL, '&nbsp;&nbsp;');
```
Comment (mattwire) [inline CRM/Admin/Form/SettingTrait.php]: Most YesNo fields are displayed with Yes as the first element (see Settings -> Search Preferences and custom fields of type YesNo).
Can we flip this array so Yes is shown first on the UI?

**Comment by seamuslee001** [comment]
@eileenmcnaughton needs rebase again now

File: CRM/Admin/Form/SettingTrait.php (hunk 1 above)
Comment (eileenmcnaughton) [reply CRM/Admin/Form/SettingTrait.php]: done

**Comment by eileenmcnaughton** [comment]
@mattwire @seamuslee001 this is now updated - note I had to change the url rule to not fail an empty value - I think that is sensible because the rule is checking value validity but the responsibility for checking presentness is elsewhere

**Comment by mattwire** [comment]
I've tested this locally and I am happy that it is working.  There is an issue with redirecting away from the page when validation fails on the URL but this was present before the changes.  Merge on pass

**Pull request merged**